### DISCUSSION TITLE
EXIF pointer tags cause other tag changes to be reset?

### DISCUSSION THREAD

**Discussion by FRbrt** [python-pillow/Pillow#6663] [Q&A] [answered]
Hello,

I'm completely new to Python / Pillow, sorry if my question isn't appropriate.
I've observed a behaviour, and I'm not sure if is that a bug, or I'm doing something wrong. Most probably I'm doing something wrong :-). I'm just very curious to find out an answer / explanation.

The situation is:
* I have a photo with EXIF data: dateTimeOriginal, GPS Info, etc.
* using the Image module I've retrieved the exif and changed the dateTimeOriginal tag

```python
with Image.open(imageFilePath) as image:
        imgExif = image.getexif()

        imgExif[0x0132] = '2022:10:05 17:00:30'   # DateTime
        imgExif[0x9003] = '2022:10:05 17:00:30'   # DateTimeOriginal
        imgExif[0x9004] = '2022:10:05 17:00:30'   # DateTimeDigitized
        image.save(outputImagePath, exif=imgExif.tobytes())
```
* after saving the file and checking the EXIF using an image viewer I see the original values for the fields above, and not the ones set by me

I've looked a bit into the _tobytes()_ function in the _Image.py_ file. I've noticed that, when any of these (0x8825 (decimal: 34853) = GPS Tag pointer, 0x8769 (decimal: 34665) = Exif Tag pointer, 0x8225) tags exists, then the EXIF is read again from the image file and added to _ifd_ return value `(ifd is TiffImagePlugin.ImageFileDirectory_v2)`. Hence, the _ifd_ dictionary will contain an entry like:

```
    ifd[34853] = { 0x9003 : '2022:09:19 15:32:27', etc. } # the original value from the image file
```

I suppose, when it is converted to bytes, it overrides the values set previously.

Please let me know if am I doing something wrong, is this intended?

Have a great day,
Robert

**Comment by radarhere** [comment, marked as answer]
I feel like you're halfway to figuring out the solution from the code.

In terms of the specification, the problem is that while DateTime is a top-level tag, DateTimeOriginal and DateTimeDigitized are not. See https://www.awaresystems.be/imaging/tiff/tifftags/exififd.html, or look at the Group column of https://exiftool.org/TagNames/EXIF.html.

It's not that your values are being overwritten as such. It's that you are setting them in the wrong place.

The following code should save those values correctly.
```python
from PIL import Image
with Image.open(imageFilePath) as image:
    imgExif = image.getexif()

    imgExif[0x0132] = '2022:10:05 17:00:30'   # DateTime

    exif_ifd = imgExif.get_ifd(0x8769)
    exif_ifd[0x9003] = '2022:10:05 17:00:30'   # DateTimeOriginal
    exif_ifd[0x9004] = '2022:10:05 17:00:30'   # DateTimeDigitized
    imgExif[0x8769] = exif_ifd
    
    image.save(outputImagePath, exif=imgExif.tobytes())
```

**Reply by FRbrt** [reply]
Works perfectly!!! Thank you very much for your quick answer!

**Reply by Routhinator** [reply]
I kept getting this obscure `required argument is not an integer` exception trying to save exif datetimeoriginal to files and this was the answer!
